### PR TITLE
Update koa

### PR DESCRIPTION
**Pull request by jkeylu**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped#common-mistakes).
- [x] Run `tsc` without errors.
- [ ] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/koajs/koa
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
Seems fine, but I'd like a sign-off from the original contributors.

**Comment by horiuchi** [comment]
LGTM 👍 
And also, could you add the test code?

File: koa/index.d.ts [hunk 1]
```diff
@@ -14,172 +14,650 @@
...
-    onerror(err: any): void;
-    use(middleware: Koa.Middleware): Koa;
+    /**
+     * Shorthand for:
+     *
+     *    http.createServer(app.callback()).listen(...)
+     */
+    listen(port?: number, hostname?: string, backlog?: number, listeningListener?: () => void): Server;
+    listen(port: number, hostname?: string, listeningListener?: () => void): Server;
+    // tslint:disable-next-line:unified-signatures
```
Comment (DanielRosenwasser) [inline koa/index.d.ts]: Can you also merge these into a

```ts
/* tslint:disable:unified-signatures */
// ....
/* tslint:enable:unified-signatures*/
```

**Comment by bowdenk7** [comment]
@jkeylu Looks like tests are failing. Could you please fix the tests and update the PR? Thanks!

**Comment by zhengbli** [comment]
Pinging @jkeylu again

**Comment by zhengbli** [comment]
👍

**Comment by didehvar** [comment]
I've just updated my @types/koa to use this version and I'm running into issues overriding the context interface.

In order to do things like `app.context.db = db;` (on 2.0.37 of @types/koa, before this PR - I believe) I was using the following:

```ts
declare module 'koa' {
  interface Context {
    db: IDatabase<any>;
    log: Logger;
  }
}
```

That doesn't work anymore as now `app.context` refers to `BaseContext` and throws the error `Property 'db' does not exist on type 'BaseContext'`. Unfortunately, changing the above code to use `BaseContext` doesn't seem to do anything.

What would be the correct way to go about this now @jkeylu? Apologies if this is in the wrong place.

**Comment by jkeylu** [comment]
@didehvar I think I should also export BaseContext, BaseRequest and BaseResponse. I will make a pull request later.

**Comment by didehvar** [comment]
It works perfectly with the new PR, thanks @jkeylu 👍

**Pull request merged**